op-upgrade: extended pause

tynes opened the pull request:
**Description**

Create a one off script that can be used to generate the upgrade JSON
files. This still needs to have the addresses for each network
added in the `addresses` mapping at the top of the script. It uses
a subset of the `op-upgrade` code.

<!-- Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md -->

tynes committed:
op-chain-ops: extended pause 2 step upgrades

The new storage setter contract with ability to set multiple
slots in a single call is deployed at [0xd81f43eDBCAcb4c29a9bA38a13Ee5d79278270cC](https://etherscan.io/address/0xd81f43edbcacb4c29a9ba38a13ee5d79278270cc#code).
It was deployed with create2 and should also be on goerli + sepolia and
ay other chain that has a superchain deployed on it.

Manual calls to `upgradeAndCall` are made. There isn't a great way to
support multiple upgrades at a time within `op-upgrade` so adopting a
"release" model where `op-upgrade` supports building a single version to
another single version makes the most sense. It is too difficult given
the current abstractions to build something that supports any upgrade to
any upgrade arbitrarily. This sort of system is certainly possible to
build but we would need to think more about the abstractions.

One thing is this does not support a reverse upgrade aka downgrade.

Each storage slot that was set as part of the upgrade to MCP is removed
using the `StorageSetter` contract. A link to the storage layout file
is included for each call to `setStorage`.

tynes committed:
op-upgrade: extended pause

Create a one off script that can be used to generate the upgrade JSON
files. This still needs to have the addresses for each network
added in the `addresses` mapping at the top of the script. It uses
a subset of the `op-upgrade` code.

tynes committed:
op-upgrade: extended pause deployments

All of the contracts for the sepolia networks are deployed.
Run the command using:

```bash
go run op-chain-ops/cmd/op-upgrade-extended-pause/main.go --help
```

It needs a deploy config to know what values to pass to initialize.
I do not believe we have deploy configs in contracts bedrock
for each of the networks.

The deploy process can be found in this notion link: https://www.notion.so/oplabs/Extended-Pause-Contract-Deployments-9a8c9547f1884c4e90e023c135288178?pvs=4

This also depends on https://github.com/ethereum-optimism/optimism/pull/8344
which zeros out values from the old upgrade to MCP. We need to update
8344 such that it zeros out all initialize storage slots

tynes committed:
op-upgrade: fix build

The call to `initialize` for building calldata isn't type safe.
This is why we should move to solidity for building this stuff
longer term.

It is able to produce a valid bundle though

```
{
  "version": "",
  "chainId": null,
  "createdAt": 0,
  "meta": {
    "createdFromSafeAddress": "",
    "createdFromOwnerAddress": "",
    "name": "",
    "description": ""
  },
  "transactions": [
    {
      "to": "0x189aBAAaa82DfC015A588A7dbaD6F13b1D3485Bc",
      "value": "0",
      "data": "0x9623609d00000000000000000000000058cc85b8d04ea49cc6dbd3cbffd00b4b8d6cb3ef000000000000000000000000c3c7e6f4ad6a593a9731a39fa883ec1999d7d87300000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000024c4d66de800000000000000000000000016fc5058f25648194471939df75cf27a2fdc48bc00000000000000000000000000000000000000000000000000000000",
      "contractMethod": {
        "inputs": [
          {
            "internalType": "address",
            "name": "_proxy",
            "type": "address"
          },
          {
            "internalType": "address",
            "name": "_implementation",
            "type": "address"
          },
          {
            "internalType": "bytes",
            "name": "_data",
            "type": "bytes"
          }
        ],
        "name": "upgradeAndCall",
        "payable": false
      },
      "contractInputsValues": {
        "_data": "0xc4d66de800000000000000000000000016fc5058f25648194471939df75cf27a2fdc48bc",
        "_implementation": "0xC3c7E6f4ad6a593a9731a39FA883eC1999d7D873",
        "_proxy": "0x58Cc85b8D04EA49cC6DBd3CbFFd00B4B8D6cb3ef"
      }
    },
    {
      "to": "0x189aBAAaa82DfC015A588A7dbaD6F13b1D3485Bc",
      "value": "0",
      "data": "0x9623609d000000000000000000000000d83e03d576d23c9aeab8cc44fa98d058d2176d1f000000000000000000000000532cad52e1f812eeb9c9a9571e07fef55993fefa00000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000024c4d66de800000000000000000000000058cc85b8d04ea49cc6dbd3cbffd00b4b8d6cb3ef00000000000000000000000000000000000000000000000000000000",
      "contractMethod": {
        "inputs": [
          {
            "internalType": "address",
            "name": "_proxy",
            "type": "address"
          },
          {
            "internalType": "address",
            "name": "_implementation",
            "type": "address"
          },
          {
            "internalType": "bytes",
            "name": "_data",
            "type": "bytes"
          }
        ],
        "name": "upgradeAndCall",
        "payable": false
      },
      "contractInputsValues": {
        "_data": "0xc4d66de800000000000000000000000058cc85b8d04ea49cc6dbd3cbffd00b4b8d6cb3ef",
        "_implementation": "0x532Cad52e1f812EEB9c9A9571E07Fef55993FEfA",
        "_proxy": "0xd83e03D576d23C9AEab8cC44Fa98d058D2176D1f"
      }
    },
    {
      "to": "0x189aBAAaa82DfC015A588A7dbaD6F13b1D3485Bc",
      "value": "0",
      "data": "0x9623609d000000000000000000000000fbb0621e0b23b5478b630bd55a5f21f67730b0f1000000000000000000000000e19c7a2c0bb32287731ea75da9b1c836815964f100000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000024c4d66de800000000000000000000000058cc85b8d04ea49cc6dbd3cbffd00b4b8d6cb3ef00000000000000000000000000000000000000000000000000000000",
      "contractMethod": {
        "inputs": [
          {
            "internalType": "address",
            "name": "_proxy",
            "type": "address"
          },
          {
            "internalType": "address",
            "name": "_implementation",
            "type": "address"
          },
          {
            "internalType": "bytes",
            "name": "_data",
            "type": "bytes"
          }
        ],
        "name": "upgradeAndCall",
        "payable": false
      },
      "contractInputsValues": {
        "_data": "0xc4d66de800000000000000000000000058cc85b8d04ea49cc6dbd3cbffd00b4b8d6cb3ef",
        "_implementation": "0xE19C7a2C0Bb32287731Ea75dA9B1C836815964F1",
        "_proxy": "0xFBb0621E0B23b5478B630BD55a5f21f67730B0F1"
      }
    },
    {
      "to": "0x189aBAAaa82DfC015A588A7dbaD6F13b1D3485Bc",
      "value": "0",
      "data": "0x9623609d00000000000000000000000016fc5058f25648194471939df75cf27a2fdc48bc000000000000000000000000592b7d3255a8037307d23c16cc8c13a9563c8ab100000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000024c4d66de8000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      "contractMethod": {
        "inputs": [
          {
            "internalType": "address",
            "name": "_proxy",
            "type": "address"
          },
          {
            "internalType": "address",
            "name": "_implementation",
            "type": "address"
          },
          {
            "internalType": "bytes",
            "name": "_data",
            "type": "bytes"
          }
        ],
        "name": "upgradeAndCall",
        "payable": false
      },
      "contractInputsValues": {
        "_data": "0xc4d66de80000000000000000000000000000000000000000000000000000000000000000",
        "_implementation": "0x592B7D3255a8037307d23C16cC8c13a9563c8Ab1",
        "_proxy": "0x16Fc5058F25648194471939df75CF27A2fdC48BC"
      }
    },
    {
      "to": "0x189aBAAaa82DfC015A588A7dbaD6F13b1D3485Bc",
      "value": "0",
      "data": "0x9623609d000000000000000000000000034edd2a225f7f429a63e0f1d2084b9e0a93b538000000000000000000000000ce77d580e0befbb1561376a722217017651b9dbf00000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000184f975e925000000000000000000000000fd1d2e729ae8eee2e146c033bf4400fe7528430100000000000000000000000000000000000000000000000000000000000000bc00000000000000000000000000000000000000000000000000000000000a6fe00000000000000000000000008f23bb38f531600e5d8fddaaec41f13fab46e98c0000000000000000000000000000000000000000000000000000000001c9c38000000000000000000000000057cacbb0d30b01eb2462e5dc940c161aff3230d30000000000000000000000000000000000000000000000000000000001312d00000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000003b9aca0000000000000000000000000000000000000000000000000000000000000f424000000000000000000000000000000000ffffffffffffffffffffffffffffffff00000000000000000000000000000000000000000000000000000000",
      "contractMethod": {
        "inputs": [
          {
            "internalType": "address",
            "name": "_proxy",
            "type": "address"
          },
          {
            "internalType": "address",
            "name": "_implementation",
            "type": "address"
          },
          {
            "internalType": "bytes",
            "name": "_data",
            "type": "bytes"
          }
        ],
        "name": "upgradeAndCall",
        "payable": false
      },
      "contractInputsValues": {
        "_data": "0xf975e925000000000000000000000000fd1d2e729ae8eee2e146c033bf4400fe7528430100000000000000000000000000000000000000000000000000000000000000bc00000000000000000000000000000000000000000000000000000000000a6fe00000000000000000000000008f23bb38f531600e5d8fddaaec41f13fab46e98c0000000000000000000000000000000000000000000000000000000001c9c38000000000000000000000000057cacbb0d30b01eb2462e5dc940c161aff3230d30000000000000000000000000000000000000000000000000000000001312d00000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000003b9aca0000000000000000000000000000000000000000000000000000000000000f424000000000000000000000000000000000ffffffffffffffffffffffffffffffff",
        "_implementation": "0xce77d580E0bEFbb1561376A722217017651B9dbF",
        "_proxy": "0x034edD2A225f7f429A63E0f1D2084B9E0A93b538"
      }
    }
  ]
}

```

refcell commented on op-chain-ops/cmd/op-upgrade-extended-pause/main.go:
```diff
@@ -0,0 +1,206 @@
+		return errors.New("Cannot create L1 client")
+	}
+	if clients.L2Client == nil {
+		return errors.New("Cannot create L1 client")
```
Rabbit looks correct here

tynes committed:
op-upgrade: fixup

tynes committed:
review: fixup

tynes committed:
op-upgrade: more cleanup

tynes committed:
op-upgrade: add in all deployments

tynes committed:
op-chain-ops: upgrades fixes

tynes committed:
contracts-bedrock: fix config

Updates the `l2OutputOracleStartingTimestamp` to a concrete value
rather than using `-1` which symbolizes "grab the latest L1 timestamp".
We often want to simplify the initial deployment but then do not want to
overwrite the value in storage when generating an upgrade transaction.
Right now `op-upgrade` uses the values in these config files. The code
enforces that it is not -1 to ensure that we do not write over the
value. The value really should be exactly the same. The state diff
tooling will be able to make assertions on this in the `superchain-ops`
post hook.

This commit simply updates the starting timestamp to the exact value
that was set on sepolia. The etherscan link can be found here:
https://sepolia.etherscan.io/address/0x90E9c4f8a994a250F6aEfd61CAFb4F2e895D458F#readProxyContract

tynes committed:
contracts-bedrock: fix base sepolia config

Makes the L1 starting blocknumber concrete.
See commit message in https://github.com/ethereum-optimism/optimism/pull/8515
for an explanation.

The value in the config comes from here:
https://sepolia.etherscan.io/address/0x84457ca9D0163FbC4bbfe4Dfbb20ba46e48DF254#readProxyContract

The contract address sourced from here:
https://github.com/base-org/contract-deployments/blob/910151eb2eed3542efbb4ccf7e38d1ed3b79d8f5/sepolia/2023-09-26-deploy/deployed/addresses.json#L10C27-L10C69

tynes committed:
contracts-bedrock: update sepolia deploy config

See https://github.com/ethereum-optimism/optimism/pull/8515 for
explanation. Also related to https://github.com/ethereum-optimism/optimism/pull/8519.

Value fetched from:
https://sepolia.etherscan.io/address/0xD5bAc3152ffC25318F848B3DD5dA6C85171BaEEe#readProxyContract

Address sourced from:
https://github.com/ethereum-optimism/superchain-registry/blob/fb3433d89a72df5809255c065b4783fa393608d8/superchain/extra/addresses/sepolia/pgn.json#L6C27-L6C69

tynes committed:
contracts-bedrock: fix zora-sepolia config

See explanation in https://github.com/ethereum-optimism/optimism/pull/8515

Related to:
- https://github.com/ethereum-optimism/optimism/pull/8519
- https://github.com/ethereum-optimism/optimism/pull/8520

Value sourced from:
https://sepolia.etherscan.io/address/0x2615B481Bd3E5A1C0C7Ca3Da1bdc663E8615Ade9#readProxyContract

Address sourced from:
https://github.com/ethereum-optimism/superchain-registry/blob/fb3433d89a72df5809255c065b4783fa393608d8/superchain/extra/addresses/sepolia/zora.json#L6C27-L6C69

Inphi commented on op-chain-ops/upgrades/l1.go:
```diff
@@ -51,12 +65,47 @@ func L1CrossDomainMessenger(batch *safe.Batch, implementations superchain.Implem
+			},
+		}
+
+		calldata, err := storageSetterABI.Pack("setBytes32", input)
```
nit: use a string constant rather than a literal for `setBytes32`

maurelian replied:
Fixed.

Inphi commented on op-chain-ops/upgrades/l1.go:
```diff
@@ -166,21 +258,58 @@ func L2OutputOracle(batch *safe.Batch, implementations superchain.Implementation
+			},
+		}
+
+		calldata, err := storageSetterABI.Pack("setBytes32", input)
+		if err != nil {
+			return err
+		}
+		args := []any{
+			common.HexToAddress(list.L2OutputOracleProxy.String()),
+			storageSetterAddr,
+			calldata,
+		}
+		proxyAdmin := common.HexToAddress(list.ProxyAdmin.String())
+		if err := batch.AddCall(proxyAdmin, common.Big0, upgradeAndCall, args, proxyAdminABI); err != nil {
+			return err
+		}
```
nit: this and elsewhere can be deduped in a function:
```go
func addStorageMigrationCall(batch *safe.Batch, list superchain.AddressList, proxy superchain.Address, proxyAdminABI *abi.ABI, migrations []bindings.StorageSetterSlot) error
```

maurelian replied:
Yes, we can/should do this in the reusable tooling. This being a 1 off I'm not concerned. :)

Inphi commented on op-chain-ops/upgrades/l1.go:
```diff
@@ -310,6 +478,76 @@ func SystemConfig(batch *safe.Batch, implementations superchain.ImplementationLi
+				Key:   common.Hash{31: 106},
+				Value: common.Hash{},
+			},
+			// bytes32 public constant L1_CROSS_DOMAIN_MESSENGER_SLOT = bytes32(uint256(keccak256("systemconfig.l1crossdomainmessenger")) - 1);
+			{
+				Key:   common.HexToHash("0x383f291819e6d54073bc9a648251d97421076bdd101933c0c022219ce9580636"),
+				Value: common.Hash{},
+			},
```
nit: really should simply compute these in Go, to minimize fat-finger errors.

maurelian replied:
I agree, we should and will improve this in the main op-upgrade tool, but for now I'm going to resolve this as 'wontfix' and I'm not worried about fat fingers because of the state diff checking processes we have in place.

Inphi reviewed (approved):
LGTM!

maurelian committed:
op-upgrade: Rename variable to more descriptive proxyAddresses

maurelian committed:
op-upgrade: DRY method string definition
